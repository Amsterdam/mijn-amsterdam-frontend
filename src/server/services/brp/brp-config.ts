export const LANDCODE_NEDERLAND = '6030';
export const LANDCODE_ONBEKEND = '0000';
export const GEMEENTE_CODE_AMSTERDAM = '0363';

/**
 * Deze waardes corresponderen met 2 verschillende onderzoekstypen:
 * 080000: Er wordt onderzocht of persoon nog op dit huidig adres verblijft.
 * 089999: Er is vastgesteld dat persoon niet meer op adres verblijft, onderzoek naar huidige verblijfplaats loopt nog.
 */
export const ADRES_IN_ONDERZOEK_A = '080000';
export const ADRES_IN_ONDERZOEK_B = '089999';
