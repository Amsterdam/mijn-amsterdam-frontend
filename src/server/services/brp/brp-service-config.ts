import { IS_PRODUCTION } from '../../../universal/config/env';

export const featureToggle = {
  router: {
    protected: {
      isEnabled: !IS_PRODUCTION, // for now only enabled in non-prod environments
    },
  },
} as const;

export const routes = {
  protected: {
    BRP_PERSONEN_RAW: `/services/brp/personen/raw`,
    BRP_VERBLIJFPLAATSHISTORIE_RAW: `/services/brp/verblijfplaatshistorie/raw`,
  },
} as const;
