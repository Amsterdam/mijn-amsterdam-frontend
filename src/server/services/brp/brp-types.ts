type TyperingSource = {
  code: string;
  omschrijving: string;
};

type NaamSource = {
  aanduidingNaamgebruik: TyperingSource;
  adellijkeTitelPredicaat: TyperingSource & { soort: 'predicaat' | 'titel' };
  voornamen: string;
  voorvoegsel: string;
  geslachtsnaam: string;
  voorletters: string;
  volledigeNaam: string;
};

type NationaliteitSource = {
  type: string;
  datumIngangGeldigheid: DatumSource;
  nationaliteit: TyperingSource;
  redenOpname: TyperingSource;
};

type GeboorteSource = {
  land: TyperingSource;
  plaats: TyperingSource;
  datum: DatumSource;
};

type OverlijdenSource = {
  datum: DatumSource;
};

type InOnderzoek = Record<string, boolean>;

export type VerblijfplaatsSource = {
  type:
    | 'Adres'
    | 'Locatie'
    | 'VerblijfplaatsBuitenland'
    | 'VerblijfplaatsOnbekend';
  verblijfadres: VerblijfadresSource;
  functieAdres: TyperingSource;
  indicatieVastgesteldVerblijftNietOpAdres: boolean;
  adresseerbaarObjectIdentificatie: string;
  nummeraanduidingIdentificatie: string;
  datumVan: DatumSource;
  inOnderzoek: InOnderzoek;
};

export type VerblijfadresSource = {
  officieleStraatnaam: string;
  korteStraatnaam: string;
  huisnummer: number;
  huisnummertoevoeging: string;
  huisletter: string;
  postcode: string;
  woonplaats: string;
  inOnderzoek: InOnderzoek;
  land?: TyperingSource; // Is only a property of VerblijfAdresBuitenland.
  locatiebeschrijving?: string; // Is only a property of VerblijfadresLocatie.
};

type ImmigratieSource = {
  landVanwaarIngeschreven: TyperingSource;
};

type AdresseringSource = {
  aanhef: string;
  aanschrijfwijze: AanschrijfwijzeSource;
  gebruikInLopendeTekst: string;
  adresregel1: string;
  adresregel2: string;
  land?: TyperingSource;
};

type AanschrijfwijzeSource = {
  naam: string;
};

type JaarDatum = {
  type: 'JaarDatum';
  jaar: string;
  langFormaat: string;
};

type JaarMaandDatum = {
  type: 'JaarMaandDatum';
  jaar: string;
  maand: string;
  langFormaat: string;
};

type VolledigeDatum = {
  type: 'Datum';
  datum: string;
  langFormaat: string;
};

type DatumOnbekend = {
  type: 'DatumOnbekend';
  onbekend: true;
  langFormaat: string;
};

export type DatumSource =
  | JaarDatum
  | JaarMaandDatum
  | VolledigeDatum
  | DatumOnbekend;

export type PersoonBasisSource = {
  naam: NaamSource;
  overlijden: OverlijdenSource;
  geboorte: GeboorteSource;
};

type KindSource = PersoonBasisSource;
type OuderSource = PersoonBasisSource;

type PartnerSource = PersoonBasisSource & {
  aangaanHuwelijkPartnerschap: HuwelijkPartnerschapSource;
  ontbindingHuwelijkPartnerschap: HuwelijkPartnerschapSource;
  soortVerbintenis: TyperingSource;
};

type HuwelijkPartnerschapSource = {
  datum: DatumSource;
  land?: TyperingSource;
  plaats?: TyperingSource;
};

type PersoonSource = PersoonBasisSource & {
  aNummer: string;
  geheimhoudingPersoonsgegevens: boolean;
  burgerservicenummer: string;
  geslacht: TyperingSource;
  leeftijd: number;
  nationaliteiten: NationaliteitSource[];
  verblijfplaats: VerblijfplaatsSource;
  immigratie: ImmigratieSource;
  gemeenteVanInschrijving: TyperingSource;
  datumInschrijvingInGemeente: DatumSource;
  adressering: AdresseringSource;
  kinderen: KindSource[];
  ouders: OuderSource[];
  partners: PartnerSource[];
};

export type PersonenResponseSource = {
  type: string;
  personen: PersoonSource[];
};

export type VerblijfplaatshistorieResponseSource = {
  verblijfplaatsen: VerblijfplaatsSource[];
};

export type Adres = {
  straatnaam: string | null;
  postcode: string | null;
  woonplaatsNaam: string | null;
  landnaam: string | null;
  huisnummer: string | null;
  huisnummertoevoeging: string | null;
  huisletter: string | null;
  begindatumVerblijf: string | null;
  begindatumVerblijfFormatted?: string | null;
  locatiebeschrijving?: string | null;
  /** @deprecated Onderstaande gegevens worden niet meer gebruikt. */
  einddatumVerblijf?: string | null;
  /** @deprecated */
  mokum?: boolean;
  /** @deprecated */
  _adresSleutel?: string;
};

export type PersoonBasis = {
  geboortedatum: string | null;
  geboortedatumFormatted?: string | null;
  overlijdensdatum?: string | null;
  overlijdensdatumFormatted?: string | null;
  geboortelandnaam: string | null;
  geboorteplaatsnaam: string | null;
  geslachtsnaam: string | null;
  omschrijvingAdellijkeTitel: string | null;
  opgemaakteNaam: string | null;
  voornamen: string | null;
  voorvoegselGeslachtsnaam: string | null;
};

export type Persoon = PersoonBasis & {
  bsn: string | null;
  gemeentenaamInschrijving: string | null;
  omschrijvingBurgerlijkeStaat: 'Ongehuwd' | null;
  omschrijvingGeslachtsaanduiding: string | null;
  nationaliteiten: Array<{ omschrijving: string }>;
  mokum: boolean;
  vertrokkenOnbekendWaarheen: boolean;
  datumVertrekUitNederland: string | null;
  datumVertrekUitNederlandFormatted?: string | null;
  indicatieGeheim: boolean;
  adresInOnderzoek: '080000' | '089999' | null;

  /** @deprecated Deze gegevens worden niet meer gebruikt. */
  aanduidingNaamgebruikOmschrijving: string | null;
  indicatieGeboortedatum?: 'J' | 'M' | 'D' | 'V' | null;
};

export type Verbintenis = {
  datumOntbinding: string | null;
  datumOntbindingFormatted?: string | null;
  datumSluiting: string | null;
  datumSluitingFormatted?: string | null;
  persoon: PersoonBasis;

  /** @deprecated Deze gegevens worden in de BENK-BRP niet meer gebruikt. */
  plaatsnaamSluitingOmschrijving?: string | null;
  soortVerbintenis?: string | null;
  soortVerbintenisOmschrijving?: string | null;
};

export type Kind = PersoonBasis;
export type Ouder = PersoonBasis;

export type BrpFrontend = {
  persoon: Persoon;
  verbintenis: Verbintenis | null;
  kinderen: Kind[];
  ouders: Ouder[];
  adres: Adres | null;
  adresHistorisch: Adres[];
};
