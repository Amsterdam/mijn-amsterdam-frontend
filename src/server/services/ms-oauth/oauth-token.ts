import qs from 'qs';

import type { ApiResponse } from '../../../universal/helpers/api';
import type { DataRequestConfig, SourceApiName } from '../../config/source-api';
import { getApiConfig } from '../../helpers/source-api-helpers';
import { requestData } from '../../helpers/source-api-request';

type TokenRequestConfigOptions = {
  sourceApiName: SourceApiName;
  tokenValidityMS: number;
  url?: string;
};

type TokenAuthOptions = {
  clientID: string;
  clientSecret: string;
  tenantID?: string;
  scope?: string;
};

type TokenPayload = {
  client_id: string;
  client_secret: string;
  grant_type: 'client_credentials';
  scope?: string;
};

type TokenResponseSource = {
  access_token: string;
  token_type: string;
};

type TokenAuthHeader = {
  Authorization: `${string} ${string}`;
};

export async function fetchAuthTokenHeader(
  requestConfigOptions: TokenRequestConfigOptions,
  authOptions: TokenAuthOptions
): Promise<ApiResponse<TokenAuthHeader>> {
  const payload: TokenPayload = {
    client_id: authOptions.clientID,
    client_secret: authOptions.clientSecret,
    grant_type: 'client_credentials',
  };
  if (authOptions.scope) {
    payload.scope = authOptions.scope;
  }
  const additionalConfig: DataRequestConfig = {
    data: qs.stringify(payload),
    cacheKey_UNSAFE: `${requestConfigOptions.sourceApiName}-oauth-access-token`, // Every request to the service (identified by serviceID) will use the same access_token so we cache it with a static key.
    cacheTimeout: requestConfigOptions.tokenValidityMS,
    transformResponse: (response: TokenResponseSource | null) => {
      if (response?.access_token) {
        return {
          Authorization: `${response.token_type} ${response.access_token}`,
        };
      }
      throw new Error(
        `${requestConfigOptions.sourceApiName}: Invalid token response`
      );
    },
    headers: {
      'Content-Type': 'application/x-www-form-urlencoded',
    },
  };

  if (!requestConfigOptions.url) {
    additionalConfig.formatUrl = (config) => {
      return `${config.url?.replaceAll(':tenant', authOptions.tenantID ?? 'common')}`;
    };
  } else {
    additionalConfig.url = requestConfigOptions.url;
  }

  const dataRequestConfig = getApiConfig('MS_OAUTH', additionalConfig);

  const tokenHeaderResponse =
    await requestData<TokenAuthHeader>(dataRequestConfig);

  return tokenHeaderResponse;
}
