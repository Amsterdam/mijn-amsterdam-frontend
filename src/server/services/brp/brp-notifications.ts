import { fetchBrpV2 } from './brp';
import { ADRES_IN_ONDERZOEK_A } from './brp-config';
import {
  themaIdBRP,
  themaTitle,
  routeConfig as routeConfigBrp,
  featureToggle,
} from '../../../client/pages/Thema/Profile/Profile-thema-config';
import {
  apiSuccessResult,
  apiDependencyError,
} from '../../../universal/helpers/api';
import { defaultDateFormat } from '../../../universal/helpers/date';
import type { MyNotification } from '../../../universal/types/App.types';
import type { AuthProfileAndToken } from '../../auth/auth-types';
import { fetchBrpNotifications } from '../profile/brp';
import type { BRPData } from '../profile/brp.types';

export function transformBRPNotifications(data: BRPData, compareDate: Date) {
  const adresInOnderzoek = data?.persoon?.adresInOnderzoek;
  const isOnbekendWaarheen = data?.persoon?.vertrokkenOnbekendWaarheen || false;
  const dateLeft = data?.persoon?.datumVertrekUitNederland
    ? defaultDateFormat(data?.persoon.datumVertrekUitNederland)
    : 'Onbekend';

  const notifications: MyNotification[] = [];

  if (adresInOnderzoek) {
    notifications.push({
      themaID: themaIdBRP,
      themaTitle: themaTitle.BRP,
      datePublished: compareDate.toISOString(),
      isAlert: true,
      id: 'brpAdresInOnderzoek',
      title: 'Adres in onderzoek',
      description:
        adresInOnderzoek === ADRES_IN_ONDERZOEK_A
          ? 'Op dit moment onderzoeken wij of u nog steeds woont op het adres waar u ingeschreven staat.'
          : 'Op dit moment onderzoeken wij op welk adres u nu woont.',
      link: {
        to: routeConfigBrp.themaPageBRP.path,
        title: 'Meer informatie',
      },
    });
  }

  if (isOnbekendWaarheen) {
    notifications.push({
      themaID: themaIdBRP,
      themaTitle: themaTitle.BRP,
      datePublished: compareDate.toISOString(),
      isAlert: true,
      id: 'brpVertrokkenOnbekendWaarheen',
      title: 'Vertrokken Onbekend Waarheen (VOW)',
      description: `U staat sinds ${dateLeft} in de Basisregistratie Personen (BRP) geregistreerd als 'vertrokken onbekend waarheen'.`,
      link: {
        to: routeConfigBrp.themaPageBRP.path,
        title: 'Meer informatie',
      },
    });
  }

  return notifications;
}

export async function fetchBrpNotificationsV2(
  authProfileAndToken: AuthProfileAndToken
) {
  if (!featureToggle[themaIdBRP].benkBrpServiceActive) {
    return fetchBrpNotifications(authProfileAndToken);
  }

  const BRP = await fetchBrpV2(authProfileAndToken);

  if (BRP.status === 'OK') {
    return apiSuccessResult({
      notifications: transformBRPNotifications(BRP.content, new Date()),
    });
  }
  return apiDependencyError({ BRP });
}
