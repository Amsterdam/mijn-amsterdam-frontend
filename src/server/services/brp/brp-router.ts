import type { Request } from 'express';

import { fetchBrpByBsn } from './brp';
import { featureToggle, routes } from './brp-service-config';
import { IS_PRODUCTION } from '../../../universal/config/env';
import {
  createBFFRouter,
  type ResponseAuthenticated,
} from '../../routing/route-helpers';

const protectedRouter = createBFFRouter({
  id: 'protected-brp',
  isEnabled: featureToggle.router.protected.isEnabled,
});

if (!IS_PRODUCTION) {
  protectedRouter.get(
    routes.protected.BRP_PERSONEN_RAW,
    async (_req: Request, res: ResponseAuthenticated) => {
      const RAW_DATA = true;
      return res.send(
        await fetchBrpByBsn(
          res.locals.sessionId,
          [res.locals.authProfileAndToken.profile.id],
          RAW_DATA
        )
      );
    }
  );
}
