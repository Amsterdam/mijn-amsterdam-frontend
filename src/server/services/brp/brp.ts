import { ONE_HOUR_MS } from '../../config/app';
import { getFromEnv } from '../../helpers/env';
import { fetchAuthTokenHeader } from '../ms-oauth/oauth-token';

const TOKEN_VALIDITY_PERIOD = 24 * ONE_HOUR_MS;
const PERCENTAGE_DISTANCE_FROM_EXPIRY = 0.1;

function fetchBenkBrpTokenHeader() {
  return fetchAuthTokenHeader(
    {
      serviceID: 'BRP',
      tokenValidityMS:
        TOKEN_VALIDITY_PERIOD * (1 - PERCENTAGE_DISTANCE_FROM_EXPIRY),
    },
    {
      clientID: getFromEnv('BFF_BENK_BRP_CLIENT_ID') ?? '',
      clientSecret: getFromEnv('BFF_BENK_BRP_CLIENT_SECRET') ?? '',
      tenantID: getFromEnv('BFF_BENK_BRP_TENANT') ?? '',
      scope: `${getFromEnv('BFF_BENK_BRP_APPLICATION_ID')}/.default`,
    }
  );
}
