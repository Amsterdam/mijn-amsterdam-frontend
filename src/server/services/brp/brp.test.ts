import { describe, it, expect, vi } from 'vitest';
import type { Mock } from 'vitest';

import { forTesting, fetchBrpByBsn } from './brp';
import testPersonenResponse from '../../../../mocks/fixtures/brp/test-personen.json';
import verblijfplaatsenResponse from '../../../../mocks/fixtures/brp/verblijfplaatshistorie.json';
import { remoteApi } from '../../../testing/utils';
import { fetchAuthTokenHeader } from '../ms-oauth/oauth-token';

const {
  fetchBenkBrpTokenHeader,
  getDatum,
  transformBenkBrpResponse,
  translateBSN,
} = forTesting;

vi.mock('../ms-oauth/oauth-token', () => ({
  fetchAuthTokenHeader: vi.fn(),
}));

describe('brp.ts', () => {
  afterAll(() => {
    vi.resetAllMocks();
    vi.restoreAllMocks();
  });

  describe('fetchBenkBrpTokenHeader', () => {
    it('should call fetchAuthTokenHeader with correct parameters', () => {
      fetchBenkBrpTokenHeader();

      expect(fetchAuthTokenHeader).toHaveBeenCalledWith(
        {
          sourceApiName: 'BRP',
          tokenValidityMS: expect.any(Number),
        },
        {
          clientID: 'test-client-id',
          clientSecret: 'test-client-secret',
          tenantID: 'test-tenant',
          scope: 'test-app-id/.default',
        }
      );
    });
  });

  describe('getDatum', () => {
    it('should return null for undefined datum', () => {
      expect(getDatum(undefined)).toBeNull();
    });

    it('should return formatted date for JaarDatum', () => {
      expect(
        getDatum({
          type: 'JaarDatum',
          jaar: '2023',
          langFormaat: '',
        })
      ).toBe('2023-01-01');
    });

    it('should return formatted date for JaarMaandDatum', () => {
      expect(
        getDatum({
          type: 'JaarMaandDatum',
          jaar: '2023',
          maand: '5',
          langFormaat: '',
        })
      ).toBe('2023-05-01');
    });

    it('should return date for Datum', () => {
      expect(
        getDatum({
          type: 'Datum',
          datum: '2023-05-15',
          langFormaat: '',
        })
      ).toBe('2023-05-15');
    });

    it('should return null for DatumOnbekend', () => {
      expect(
        getDatum({
          type: 'DatumOnbekend',
          langFormaat: '',
          onbekend: true,
        })
      ).toBeNull();
    });
  });

  describe('transformBenkBrpResponse', () => {
    it('should throw an error if no person is found in response', () => {
      expect(() =>
        transformBenkBrpResponse({
          personen: [],
          type: '',
        })
      ).toThrow('No person found in Benk BRP response');
    });

    it('should transform response data correctly', () => {
      const responseData = {
        personen: [
          {
            naam: { volledigeNaam: 'John Doe' },
            verblijfplaats: { type: 'VerblijfplaatsOnbekend' },
            gemeenteVanInschrijving: { code: '0363' },
          },
        ],
      };

      const result = transformBenkBrpResponse(responseData as any);
      expect(result).toHaveProperty('persoon.opgemaakteNaam', 'John Doe');
      expect(result).toHaveProperty('persoon.vertrokkenOnbekendWaarheen', true);
      expect(result).toHaveProperty('persoon.mokum', true);
    });
  });

  describe('translateBSN', () => {
    it('should return the same BSN if translations are not set', () => {
      expect(translateBSN('123456789')).toBe('123456789');
    });

    it('should translate BSN if translations are set', () => {
      process.env.BFF_BENK_BSN_TRANSLATIONS =
        '123456789=987654321,111111111=222222222';
      expect(translateBSN('123456789')).toBe('987654321');
      expect(translateBSN('111111111')).toBe('222222222');
    });
  });

  describe('fetchBrpByBsn', () => {
    it('should return response if fetchAuthTokenHeader fails', async () => {
      (fetchAuthTokenHeader as Mock).mockResolvedValue({ status: 'ERROR' });

      const result = await fetchBrpByBsn('test-session-id', ['123456789']);
      expect(result).toEqual({ status: 'ERROR' });
    });

    describe('Should format response data correctly', () => {
      test('Formatting for test BSN 1, with verblijfplaatshistorie', async () => {
        vi.unmock('../../helpers/source-api-request');

        (fetchAuthTokenHeader as Mock).mockResolvedValue({
          content: { Authorization: 'Bearer test-token' },
          status: 'OK',
        });

        const BSN = '999971074';
        const testPersoon = testPersonenResponse.personen.find(
          (p) => p.burgerservicenummer === BSN
        );
        remoteApi.post(/\/personen/).reply(200, { personen: [testPersoon] });
        remoteApi
          .post(/\/verblijfplaatshistorie/)
          .reply(200, verblijfplaatsenResponse);

        const response = await fetchBrpByBsn('test-session-id', [BSN]);

        expect(response).toMatchSnapshot();
        expect(response.status).toBe('OK');
        expect(response.content?.persoon.bsn).toBe(BSN);
      });

      test('Formatting for test BSN 2, with verblijfplaatshistorie returning an error', async () => {
        vi.unmock('../../helpers/source-api-request');

        (fetchAuthTokenHeader as Mock).mockResolvedValue({
          content: { Authorization: 'Bearer test-token' },
          status: 'OK',
        });

        const BSN = '999990810';
        const testPersoon = testPersonenResponse.personen.find(
          (p) => p.burgerservicenummer === BSN
        );
        remoteApi.post(/\/personen/).reply(200, { personen: [testPersoon] });
        remoteApi
          .post(/\/verblijfplaatshistorie/)
          .reply(500, 'Internal Server Error');

        const response = await fetchBrpByBsn('test-session-id', [BSN]);

        expect(response).toMatchSnapshot();
        expect(response.status).toBe('OK');
        expect(response.content?.persoon.bsn).toBe(BSN);
      });
    });
  });
});
