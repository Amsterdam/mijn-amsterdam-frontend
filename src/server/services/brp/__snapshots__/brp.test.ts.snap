// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`brp.ts > fetchBrpByBsn > Should format response data correctly > Formatting for test BSN 1, with verblijfplaatshistorie 1`] = `
{
  "content": {
    "adres": {
      "begindatumVerblijf": "2015-10-01",
      "begindatumVerblijfFormatted": "1 oktober 2015",
      "huisletter": null,
      "huisnummer": "1",
      "huisnummertoevoeging": null,
      "landnaam": null,
      "locatiebeschrijving": null,
      "postcode": "1102BL",
      "straatnaam": "E Cady Stantonpln",
      "woonplaatsNaam": "Amsterdam",
    },
    "adresHistorisch": [
      {
        "begindatumVerblijf": "1995-10-20",
        "begindatumVerblijfFormatted": "20 oktober 1995",
        "huisletter": null,
        "huisnummer": "4",
        "huisnummertoevoeging": null,
        "landnaam": null,
        "locatiebeschrijving": null,
        "postcode": "1098NK",
        "straatnaam": "Ampèrestraat",
        "woonplaatsNaam": "Amsterdam",
      },
      {
        "begindatumVerblijf": "1995-04-12",
        "begindatumVerblijfFormatted": "12 april 1995",
        "huisletter": null,
        "huisnummer": "100",
        "huisnummertoevoeging": null,
        "landnaam": null,
        "locatiebeschrijving": null,
        "postcode": "1011AH",
        "straatnaam": "Prins Hendrikkade",
        "woonplaatsNaam": null,
      },
    ],
    "kinderen": [
      {
        "geboortedatum": "2015-09-01",
        "geboortedatumFormatted": "1 september 2015",
        "geboortelandnaam": "Nigeria",
        "geboorteplaatsnaam": "Lagos",
        "geslachtsnaam": "Abayomi - Akwasi",
        "omschrijvingAdellijkeTitel": null,
        "opgemaakteNaam": null,
        "overlijdensdatum": null,
        "overlijdensdatumFormatted": null,
        "voornamen": null,
        "voorvoegselGeslachtsnaam": null,
      },
    ],
    "ouders": [
      {
        "geboortedatum": "1979-11-12",
        "geboortedatumFormatted": "12 november 1979",
        "geboortelandnaam": "Nigeria",
        "geboorteplaatsnaam": "Lagos",
        "geslachtsnaam": "Bawana",
        "omschrijvingAdellijkeTitel": null,
        "opgemaakteNaam": null,
        "overlijdensdatum": null,
        "overlijdensdatumFormatted": null,
        "voornamen": null,
        "voorvoegselGeslachtsnaam": null,
      },
      {
        "geboortedatum": "1975-06-06",
        "geboortedatumFormatted": "6 juni 1975",
        "geboortelandnaam": "Nigeria",
        "geboorteplaatsnaam": "Lagos",
        "geslachtsnaam": "Akwasi",
        "omschrijvingAdellijkeTitel": null,
        "opgemaakteNaam": null,
        "overlijdensdatum": null,
        "overlijdensdatumFormatted": null,
        "voornamen": null,
        "voorvoegselGeslachtsnaam": null,
      },
    ],
    "persoon": {
      "aanduidingNaamgebruikOmschrijving": "eigen geslachtsnaam",
      "adresInOnderzoek": null,
      "bsn": "999971074",
      "datumVertrekUitNederland": null,
      "datumVertrekUitNederlandFormatted": null,
      "geboortedatum": "1995-09-01",
      "geboortedatumFormatted": "1 september 1995",
      "geboortelandnaam": "Nigeria",
      "geboorteplaatsnaam": "Lagos",
      "gemeentenaamInschrijving": "Amsterdam",
      "geslachtsnaam": "Akwasi",
      "indicatieGeheim": false,
      "mokum": true,
      "nationaliteiten": [],
      "omschrijvingAdellijkeTitel": null,
      "omschrijvingBurgerlijkeStaat": "Ongehuwd",
      "omschrijvingGeslachtsaanduiding": "vrouw",
      "opgemaakteNaam": "Akwasi",
      "overlijdensdatum": null,
      "overlijdensdatumFormatted": null,
      "vertrokkenOnbekendWaarheen": false,
      "voornamen": null,
      "voorvoegselGeslachtsnaam": null,
    },
    "verbintenis": {
      "datumOntbinding": "2019-11-05",
      "datumOntbindingFormatted": "5 november 2019",
      "datumSluiting": "2014-08-01",
      "datumSluitingFormatted": "1 augustus 2014",
      "persoon": {
        "geboortedatum": "1989-11-12",
        "geboortedatumFormatted": "12 november 1989",
        "geboortelandnaam": "Nigeria",
        "geboorteplaatsnaam": "Lagos",
        "geslachtsnaam": "Abayomi",
        "omschrijvingAdellijkeTitel": null,
        "opgemaakteNaam": null,
        "overlijdensdatum": null,
        "overlijdensdatumFormatted": null,
        "voornamen": null,
        "voorvoegselGeslachtsnaam": null,
      },
    },
  },
  "status": "OK",
}
`;

exports[`brp.ts > fetchBrpByBsn > Should format response data correctly > Formatting for test BSN 2, with verblijfplaatshistorie returning an error 1`] = `
{
  "content": {
    "adres": {
      "begindatumVerblijf": "2001-05-01",
      "begindatumVerblijfFormatted": "1 mei 2001",
      "huisletter": null,
      "huisnummer": "650",
      "huisnummertoevoeging": null,
      "landnaam": null,
      "locatiebeschrijving": null,
      "postcode": "1098XH",
      "straatnaam": "Science Park",
      "woonplaatsNaam": "Amsterdam",
    },
    "adresHistorisch": [],
    "kinderen": [],
    "ouders": [
      {
        "geboortedatum": "1914-01-11",
        "geboortedatumFormatted": "11 januari 1914",
        "geboortelandnaam": "Nederland",
        "geboorteplaatsnaam": "Heiloo",
        "geslachtsnaam": "Manderley",
        "omschrijvingAdellijkeTitel": null,
        "opgemaakteNaam": null,
        "overlijdensdatum": null,
        "overlijdensdatumFormatted": null,
        "voornamen": "Johanna",
        "voorvoegselGeslachtsnaam": null,
      },
      {
        "geboortedatum": "1919-04-29",
        "geboortedatumFormatted": "29 april 1919",
        "geboortelandnaam": "Nederland",
        "geboorteplaatsnaam": "Heiloo",
        "geslachtsnaam": "Roos",
        "omschrijvingAdellijkeTitel": null,
        "opgemaakteNaam": null,
        "overlijdensdatum": null,
        "overlijdensdatumFormatted": null,
        "voornamen": "Hendrikus",
        "voorvoegselGeslachtsnaam": "de",
      },
    ],
    "persoon": {
      "aanduidingNaamgebruikOmschrijving": "eigen geslachtsnaam",
      "adresInOnderzoek": null,
      "bsn": "999990810",
      "datumVertrekUitNederland": null,
      "datumVertrekUitNederlandFormatted": null,
      "geboortedatum": "1955-06-02",
      "geboortedatumFormatted": "2 juni 1955",
      "geboortelandnaam": "Nederland",
      "geboorteplaatsnaam": "Haarlem",
      "gemeentenaamInschrijving": "Amsterdam",
      "geslachtsnaam": "Roos",
      "indicatieGeheim": false,
      "mokum": true,
      "nationaliteiten": [
        {
          "omschrijving": "Nederlandse",
        },
      ],
      "omschrijvingAdellijkeTitel": null,
      "omschrijvingBurgerlijkeStaat": null,
      "omschrijvingGeslachtsaanduiding": "vrouw",
      "opgemaakteNaam": "Hendrika Isabella Gerarda Johanna de Roos",
      "overlijdensdatum": null,
      "overlijdensdatumFormatted": null,
      "vertrokkenOnbekendWaarheen": false,
      "voornamen": "Hendrika Isabella Gerarda Johanna",
      "voorvoegselGeslachtsnaam": "de",
    },
    "verbintenis": {
      "datumOntbinding": null,
      "datumOntbindingFormatted": null,
      "datumSluiting": "2002-02-02",
      "datumSluitingFormatted": "2 februari 2002",
      "persoon": {
        "geboortedatum": "1970-02-28",
        "geboortedatumFormatted": "28 februari 1970",
        "geboortelandnaam": "Bolivia",
        "geboorteplaatsnaam": "Cochabambo",
        "geslachtsnaam": "Bruchs",
        "omschrijvingAdellijkeTitel": null,
        "opgemaakteNaam": null,
        "overlijdensdatum": null,
        "overlijdensdatumFormatted": null,
        "voornamen": "Helène",
        "voorvoegselGeslachtsnaam": null,
      },
    },
  },
  "failedDependencies": {
    "adresHistorisch": {
      "content": null,
      "message": "Unexpected token 'I', "Internal S"... is not valid JSON",
      "status": "ERROR",
    },
  },
  "status": "OK",
}
`;
